transmart-rest-api/grails-app/services/org/transmartproject/rest/SurveyTableViewDataSerializationService.groovy

```diff
@@ -32,31 +48,156 @@ import org.transmartproject.rest.serialization.tabular.TabularResultTSVSerialize
 
 import java.util.zip.ZipOutputStream
 
+import static groovyx.gpars.GParsPool.withPool
+
 @Transactional
 class SurveyTableViewDataSerializationService implements DataSerializer {
 
     @Autowired
     MultiDimensionalDataResource multiDimService
 
-    Map<Format, TabularResultSerializer> formatToSerializer = [
-            (Format.TSV) : new TabularResultTSVSerializer(),
-            (Format.SPSS): new TabularResultSPSSSerializer()
-    ]
-            .withDefault { Format format -> throw new UnsupportedOperationException("Unsupported format for tabular data: ${format}") }
+    @Autowired
+    MDStudiesResource studiesResource
+
+    @Autowired
+    ConceptsResource conceptsResource
+
+    @Autowired
+    SurveyTableColumnService surveyTableColumnService
+
+    @Autowired
+    SessionFactory sessionFactory
+
+    @Autowired
+    SystemResource systemResource
+
+    Set<Format> supportedFormats = [Format.TSV, Format.SPSS]
+
+    TabularResultSerializer getSerializer(Format format, User user, ZipOutputStream zipOutputStream, ImmutableList<DataColumn> columns) {
+        switch(format) {
+            case Format.TSV:
+                return new TabularResultTSVSerializer(user, zipOutputStream, columns)
+            case Format.SPSS:
+                return new TabularResultSPSSSerializer(user, zipOutputStream, columns)
+            default:
+                throw new UnsupportedOperationException("Unsupported format for tabular data: ${format}")
+        }
+    }
+
 
+    /**
+     *
+     * @param format
+     * @param constraint
+     * @param user
+     * @param out
+     */
     @Override
     void writeClinical(Format format,
                        MultiDimConstraint constraint,
                        User user,
                        OutputStream out) {
+        log.info "Start parallel export ..."
+        PatientSetConstraint patientSetConstraint = null
+        Constraint columnConstraint = new TrueConstraint()
+        if (constraint instanceof PatientSetConstraint) {
```

Comment:
Could you try to get constraint optimization code in a separate class? It looks like it could be brought to the rest of such code in `SurveyTableColumnService` class at least. Also the constraint optimization code looks pretty massive. We should cover this code with some unit tests.